packages/loopring_v3/contracts/impl/agents/FastWithdrawalAgent.sol

```diff
@@ -0,0 +1,247 @@
+import "../../lib/SignatureUtil.sol";
+import "../../lib/EIP712.sol";
+
+
+/// @title Fast withdrawal agent implementation. The fast withdrawal request reduces to
+///        a normal onchain withdrawal request after a specified time limit has exceeded.
+///
+///        Fast withdrawals are a way for an operator to provide instant withdrawals for
+///        users with the help of a liquidity provider and conditional transfers.
+///
+///        A fast withdrawal requires the non-trustless cooperation of 2 parties:
+///        - A liquidity provider which provides funds to users immediately onchain
+///        - The operator which will make sure the user has sufficient funds offchain
+///          so that the liquidity provider can be paid back offchain using a conditional transfer.
+///          The operator also needs to process those conditional transfers so that the
+///          liquidity provider receives its funds back in its own account where it
+///          again has full custody over it.
+///
+///        However, there is a special case when the fast withdrawal reduces to a standard
+///        withdrawal and the fee is paid onchain. In this case the withdrawal can be
+///        done completely trustless, no cooperation with the operator is needed.
+///
+///        We require the fast withdrawals to be executed by the liquidity provider (as msg.sender)
+///        so that the liquidity provider can impose its own rules on how its funds are spent. This will
+///        inevitably need to be done in close cooperation with the operator, or by the operator
+///        itself using a smart contract where the liquidity provider enforces who, how
+///        and even if their funds can be used to facilitate the fast withdrawals.
+///
+///        The liquidity provider can call `executeFastWithdrawals` to provide users
+///        immediately with funds onchain. This allows the security of the funds to be handled
+///        by any EOA or smart contract.
+///
+///        Users that want to make use of this functionality have to
+///        authorize this contract as their agent.
+///
+/// @author Brecht Devos - <brecht@loopring.org>
+contract FastWithdrawalAgent is ReentrancyGuard
+{
+    using AddressUtil       for address;
+    using AddressUtil       for address payable;
+    using ERC20SafeTransfer for address;
+    using MathUint          for uint;
+    using SignatureUtil     for bytes32;
+
+    struct FastWithdrawal
+    {
+        address exchange;
+        address from;                   // The owner of the account
+        address to;                     // The address that will receive the tokens withdrawn
+        address token;
+        uint    amount;
+        address feeToken;
+        uint    fastFee;                // Fee paid for a fast withdrawal
+        uint    fee;                    // Fee paid for a standard onchain withdrawal
+        bool    onchainFeePayment;
+        uint    fastUntil;
+        uint    validUntil;
```

Comment:
`fastUntil` is the time the request can be used for a fast withdrawal.
`fastUntil` is the time the request can be used since `fastUntil` for a normal onchain withdrawal. Otherwise the normal onchain withdrawal would be possible with the request indefinitely well after the user requested the withdrawal (and he could have done it himself in the meantime, so he would pay a fee for nothing).